### ISSUE TITLE
static-location.js is now incompatible with Jest

### ISSUE THREAD

**Issue by davidje13** [molefrog/wouter#126]
In #113, I made `static-location.js` into more than a single line of code, and in doing so I changed `module.exports =` to `export default`, to fit with the style used in `index.js` and `use-location.js`.

Unfortunately, it seems that because `static-location` is typically for testing, this has revealed an incompatibility with Jest.

I'm not sure what the correct thing to do here is. Wouter explicitly ships in es6 _and_ cjs flavours, with the cjs `index.js` marked as `main` in `package.json`, and the es6 index.js marked as `module`. BUT: `module` only applies to the main entrypoint; when including something else directly (i.e. `import staticLocationHook from 'wouter/static-location'`), that field does nothing and the file is found by path relative to root. Which is the es6 version due to the build layout used. Jest doesn't support es6 (even though the NodeJS runtime can), and everything explodes:

> ```
>   ● Test suite failed to run
> 
>     Jest encountered an unexpected token
> 
>     This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.
> 
>     By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".
> 
>     Here's what you can do:
>      • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
>      • If you need a custom transformation specify a "transform" option in your config.
>      • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.
> 
>     You'll find more details and examples of these config options in the docs:
>     https://jestjs.io/docs/en/configuration.html
> 
>     Details:
> 
>     /<project>/node_modules/wouter/static-location.js:4
>     export default (path = "/", { record = false } = {}) => {
>     ^^^^^^
> 
>     SyntaxError: Unexpected token 'export'
> ```

This could be worked around by switching it back to use `module.exports`, but then what's the point of providing both es6 and cjs versions? and should the same be done to `use-location` as well even though nobody seems to have complained about that one yet?

Another option would be to put the built cjs files in the root and have an explicit directory for es6, but that just reverses the problem (fine for people using non-module-aware tooling like Jest, but no good for anyone hoping to use the latest-and-greatest)

I've been looking around but it doesn't look like `package.json` lets you specify separate base directories for es6 vs cjs, although there are some notes about using `.js` vs `.mjs` rather than using separate directories (but I don't know which tools actually support that for discovery).

Finally, users could work around this by explicitly importing from `wouter/cjs/static-location`, but that's a bit clunky and the TypeScript definitions don't cover it.

Any ideas?

**Comment by molefrog** [comment]
Hi @davidje13, thanks for reporting this. 
Well, the subpath export support in Node is quite a common problem indeed. I have a hope that we can use a special `exports` field in package.json which comes along with ESM support in Node.js. In Node 12 (LTS) the ESM support requires no flag, so hopefully we can define proper CJS/ESM exports and then everyone who use Node >= 12, will be able to do proper require. For all clients below 12, the only proper way is to use `wouter/cjs/static-location`, but I think that's fine.

I'm planning to conduct a way experiments to see if the `exports` thing works and will get back to you soon.

**Comment by molefrog** [comment]
Ok, I tested the following with the latest Node and it worked. I was able to import `wouter/static-location` from both CJS and ESM modules. I added these lines to package.json:

```json
  "type": "module",
  "exports": {
    ".": {
      "require": "./cjs/index.js",
      "default": "./index.js"
    },
    "./static-location": {
      "require": "./cjs/static-location.js",
      "default": "./static-location.js"
    }
  }
```
It basically means that the package is ESM by default, unless it is used within `require()`, then the CJS version will be used.

I also had to create a `cjs/package.json` file with `{ "type": "commonjs" }` — this was needed because even though a CJS version is listed in `require`, the Node assumes it is an ESM module, since the type is `module` in the root package.json. An alternative is to list it with explicit `./cjs/index.cjs` extension, but I'm afraid it will break the bundlers. 

**However, there is another problem**, which is related to the fact that we build CJS with `--exports named` Rollup flag, so the output CJS file doesn't have `module.exports = `, but instead has a `default` property, so the Node users will have to write:

```js
const staticLocation = require("wouter/static-location").default
```

**Comment by davidje13** [comment]
Changing it back is certainly the quickest fix and doesn't leave things any worse than they were before. It's just a bit weird to have es6 and cjs flavours which contain exactly the same code! (and don't forget that it also technically applies to `use-location.js` too, even though nobody seems to have hit problems there yet)

I agree with a quick fix to resolve the regression, but I also think there's some more to consider here longer-term.

**Comment by molefrog** [comment]
@davidje13 Made a PR https://github.com/molefrog/wouter/pull/127 that addresses all the problems discussed in this thread. Decided not to restore `module.exports = ` because it will break ESM imports.

**Comment by molefrog** [comment]
@davidje13 Published wouter@2.5.1 with proper CJS exports.

**Comment by davidje13** [comment]
Thanks for patching it. Just wanted to note that although this fix is technically correct, it doesn't work with Jest _yet_. I think that's because the issue you linked to (https://github.com/webpack/webpack/issues/9509) is being included in WebPack 5, which hasn't been released (but hopefully will be soon).

I don't think this project should change because of that; better to wait for them to release the proper fix. In the meantime, it's easy for users to patch with the following Jest config, which is what I'm now doing:

```json
"moduleNameMapper": {
  "^wouter/(.*)$": "wouter/cjs/$1"
}
```

(https://jestjs.io/docs/en/configuration#modulenamemapper-objectstring-string--arraystring)

When users update to WebPack 5 they can remove that setting.
